model/lifecycle.go

```diff
@@ -969,3 +969,105 @@ func sortLayer(layer []task.Task, idToDisplayName map[string]string) []task.Task
 	}
 	return sortedLayer
 }
+
+// Given a patch version and a list of variant/task pairs, creates the set of new builds that
+// do not exist yet out of the set of pairs. No tasks are added for builds which already exist
+// (see AddNewTasksForPatch).
+func AddNewBuilds(activated bool, v *version.Version, p *Project, tasks TaskVariantPairs) error {
+	taskIds := NewPatchTaskIdTable(p, v, tasks)
+
+	newBuildIds := make([]string, 0)
+	newBuildStatuses := make([]version.BuildStatus, 0)
+
+	existingBuilds, err := build.Find(build.ByVersion(v.Id).WithFields(build.BuildVariantKey, build.IdKey))
+	if err != nil {
+		return err
+	}
+	variantsProcessed := map[string]bool{}
+	for _, b := range existingBuilds {
+		variantsProcessed[b.BuildVariant] = true
+	}
+
+	for _, pair := range tasks.ExecTasks {
+		if _, ok := variantsProcessed[pair.Variant]; ok { // skip variant that was already processed
+			continue
+		}
+		variantsProcessed[pair.Variant] = true
+		// Extract the unique set of task names for the variant we're about to create
+		taskNames := tasks.ExecTasks.TaskNames(pair.Variant)
+		displayNames := tasks.DisplayTasks.TaskNames(pair.Variant)
+		buildId, err := CreateBuildFromVersion(p, v, taskIds, pair.Variant, activated, taskNames, displayNames)
+		grip.Infof("Creating build for version %s, buildVariant %s, activated=%t",
+			v.Id, pair.Variant, activated)
+		if err != nil {
+			return err
+		}
+		newBuildIds = append(newBuildIds, buildId)
+		newBuildStatuses = append(newBuildStatuses,
+			version.BuildStatus{
+				BuildVariant: pair.Variant,
+				BuildId:      buildId,
+				Activated:    activated,
+			},
+		)
+	}
+
+	return version.UpdateOne(
+		bson.M{version.IdKey: v.Id},
+		bson.M{
+			"$push": bson.M{
+				version.BuildIdsKey:      bson.M{"$each": newBuildIds},
+				version.BuildVariantsKey: bson.M{"$each": newBuildStatuses},
+			},
+		},
+	)
+}
+
+// Given a version and set of variant/task pairs, creates any tasks that don't exist yet,
+// within the set of already existing builds.
+func AddNewTasks(activated bool, v *version.Version, p *Project, pairs TaskVariantPairs) error {
+	builds, err := build.Find(build.ByIds(v.BuildIds).WithFields(build.IdKey, build.BuildVariantKey, build.CreateTimeKey))
+	if err != nil {
+		return err
```

Comment:
wrap errors and make sure that you don't have to make sure that builds isn't nil/empty